<h1>Cosmetics Shop 2019 Q4 用户行为分析<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#分析目的" data-toc-modified-id="分析目的-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>分析目的</a></span></li><li><span><a href="#主要结论前置" data-toc-modified-id="主要结论前置-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>主要结论前置</a></span></li><li><span><a href="#季度整体情况" data-toc-modified-id="季度整体情况-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>季度整体情况</a></span><ul class="toc-item"><li><span><a href="#季度指标汇总" data-toc-modified-id="季度指标汇总-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>季度指标汇总</a></span><ul class="toc-item"><li><span><a href="#流量指标" data-toc-modified-id="流量指标-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>流量指标</a></span></li><li><span><a href="#行为指标" data-toc-modified-id="行为指标-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>行为指标</a></span></li><li><span><a href="#交易指标" data-toc-modified-id="交易指标-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>交易指标</a></span></li></ul></li><li><span><a href="#整体行为转化" data-toc-modified-id="整体行为转化-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>整体行为转化</a></span></li></ul></li><li><span><a href="#时间维度分析" data-toc-modified-id="时间维度分析-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>时间维度分析</a></span><ul class="toc-item"><li><span><a href="#按月分析" data-toc-modified-id="按月分析-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>按月分析</a></span></li><li><span><a href="#按周分析" data-toc-modified-id="按周分析-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>按周分析</a></span></li><li><span><a href="#按星期分析" data-toc-modified-id="按星期分析-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>按星期分析</a></span></li><li><span><a href="#按日期分析" data-toc-modified-id="按日期分析-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>按日期分析</a></span></li><li><span><a href="#按时段分析" data-toc-modified-id="按时段分析-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>按时段分析</a></span></li></ul></li><li><span><a href="#用户分析" data-toc-modified-id="用户分析-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>用户分析</a></span><ul class="toc-item"><li><span><a href="#用户行为统计信息" data-toc-modified-id="用户行为统计信息-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>用户行为统计信息</a></span></li><li><span><a href="#用户活跃情况" data-toc-modified-id="用户活跃情况-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>用户活跃情况</a></span></li><li><span><a href="#用户行为路径分布" data-toc-modified-id="用户行为路径分布-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>用户行为路径分布</a></span></li></ul></li><li><span><a href="#客户分析" data-toc-modified-id="客户分析-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>客户分析</a></span><ul class="toc-item"><li><span><a href="#用户购买情况" data-toc-modified-id="用户购买情况-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>用户购买情况</a></span></li><li><span><a href="#每日客户复购" data-toc-modified-id="每日客户复购-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>每日客户复购</a></span></li><li><span><a href="#购买次数分布" data-toc-modified-id="购买次数分布-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>购买次数分布</a></span></li><li><span><a href="#复购间隔天数分布" data-toc-modified-id="复购间隔天数分布-6.4"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>复购间隔天数分布</a></span></li><li><span><a href="#消费金额分布" data-toc-modified-id="消费金额分布-6.5"><span class="toc-item-num">6.5&nbsp;&nbsp;</span>消费金额分布</a></span></li><li><span><a href="#客户分类" data-toc-modified-id="客户分类-6.6"><span class="toc-item-num">6.6&nbsp;&nbsp;</span>客户分类</a></span><ul class="toc-item"><li><span><a href="#传统-RFM-模型" data-toc-modified-id="传统-RFM-模型-6.6.1"><span class="toc-item-num">6.6.1&nbsp;&nbsp;</span>传统 RFM 模型</a></span></li><li><span><a href="#KMeans-聚类" data-toc-modified-id="KMeans-聚类-6.6.2"><span class="toc-item-num">6.6.2&nbsp;&nbsp;</span>KMeans 聚类</a></span></li></ul></li></ul></li><li><span><a href="#品牌分析" data-toc-modified-id="品牌分析-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>品牌分析</a></span><ul class="toc-item"><li><span><a href="#品牌商品数量" data-toc-modified-id="品牌商品数量-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>品牌商品数量</a></span></li><li><span><a href="#品牌成交额" data-toc-modified-id="品牌成交额-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>品牌成交额</a></span></li><li><span><a href="#品牌矩阵分析" data-toc-modified-id="品牌矩阵分析-7.3"><span class="toc-item-num">7.3&nbsp;&nbsp;</span>品牌矩阵分析</a></span></li></ul></li></ul></div>

In [1]:
import numpy as np
import pandas as pd
pd.set_option('display.float_format',lambda x : '%.6f' % x)
pd.set_option("max_rows", 100)

import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('whitegrid')
from pyecharts.charts import *
from pyecharts.components import Table
from pyecharts.commons.utils import JsCode
import pyecharts.options as opts
from pyecharts.globals import ThemeType

from sklearn.preprocessing import MinMaxScaler
from sklearn.cluster import KMeans
from sklearn import metrics

import os
import gc
import warnings
warnings.filterwarnings('ignore')


DATA_PATH = '../data/'
CLEAN_PATH = '../clean/'
CACHE_PATH = '../cache/'
DATA_FILES = ['2019-Oct.csv', '2019-Nov.csv', '2019-Dec.csv',]
cleaned_target = f"{CLEAN_PATH}processed_data.h5"

data = pd.read_hdf(cleaned_target, key='q4')
data = pd.concat([data, pd.get_dummies(data['event_type'])], axis=1)
data['turnovers'] = data['price'] * data['purchase']
data['date'] = data['event_time'].dt.date
data['month'] = data['event_time'].dt.month
data['week'] = data['event_time'].dt.week
data['weekday'] = data['event_time'].dt.weekday
weekday_mapper = ['星期一', '星期二', '星期三', '星期四', '星期五', '星期六', '星期日', ]
weekday_mapper = dict(zip([i for i in range(len(weekday_mapper))], weekday_mapper))
data['weekday'] = data['weekday'].map(weekday_mapper)
data['hour'] = data['event_time'].dt.hour
data = data.drop(columns=['event_time'])
# data

*Author: json.wong.work@gmail.com*

*Index: https://nbviewer.org/*

## 分析目的

对季度运营数据进行探索性分析，按不同时间粒度进行汇总分析，用户分析，品牌分析，挖掘规律，寻找问题。

## 主要结论前置

1. [整体购买率仅 7.39%，新用户的运营有改善空间；复购率高达 89.20%，老客户的粘性很高。](#交易指标)
2. [黑五是用户季度活跃和成交的最高峰，圣诞节过后用户活跃和成交持续走低。](#按日期分析)
3. [活动期间，除了流量、销量，用户客单价也会明显提高。](#按日期分析)
4. [11 时和 19 时左右是每日用户活跃与成交的高峰期，客单价高于其它时段，但各环节转化率没有提升。](#按时段分析)
5. [用户浏览与重复浏览的转化率低，商品的首次曝光和重复曝光策略应该进行调整。](#用户行为路径分布)
6. [提高用户日常粘性，以期提高单个用户的 LTV](#用户活跃情况)
7. [加购后购买的转化率明显低于没有后续操作和取消购物车，推测用户“加购”不一定代表较高的购买欲望，也可能是收藏商品或价格观察。](#用户行为路径分布)
8. [客户在首日购买的商品数量（平均≥5件）明显多于复购日购买的商品数量（平均≤2件）。](#用户购买情况)
9. [购后的间隔天数越小，复购的概率越高。](#复购间隔天数分布)
10. [客户质量两极分化，一般客户基数较大有待开发。](#客户分类)
11. [对“最近购买时间”较远的用户群体应及时进行召回。](#客户分类) 
12. [高流量不一定高销量，但低流量几乎必然低销量。在高流量的情况下，低价商品销量更高，但低价并非高销量的必然前提。](#品牌矩阵分析)

## 季度整体情况

### 季度指标汇总

#### 流量指标

指标口径：
- pv：浏览量，‘view’次数
- uv：独立访客数
- 跳失率：仅有 1 次浏览行为的用户比例
- 人均浏览次数：平均每个用户浏览次数

In [2]:
pv = data['view'].sum()
uv = data['user_id'].nunique()
user_behaviour_gb = data[['user_id', 'view', 'cart', 'remove_from_cart', 'purchase']].groupby('user_id').sum()
bounce_rate = len(user_behaviour_gb[(user_behaviour_gb['view'] == 1) & \
                                    (user_behaviour_gb['cart'] + user_behaviour_gb['remove_from_cart'] + user_behaviour_gb['purchase'] == 0)]) / len(user_behaviour_gb)
per_capita_view = user_behaviour_gb['view'].mean()

headers = ['PV', 'UV', '跳失率', '人均浏览次数']
flow= [[pv, uv, f"{bounce_rate:.2%}", f"{per_capita_view:.2f}"]]
flow_table = (
    Table()
    .add(headers, flow)
)
flow_table.render_notebook()

PV,UV,跳失率,人均浏览次数
5256601,969369,45.06%,5.42


#### 行为指标

指标口径：

- 加购次数：'cart'次数
- 移出购物车次数：'remove_from_cart'次数
- 购买次数：'purchase'次数
- 平均使用天数：每日登陆用户数（去重）之和与季度登陆用户数（去重）的比值

In [3]:
cart_n, rmc_n, purchase_n = data['cart'].sum(), data['remove_from_cart'].sum(), data['purchase'].sum()
month_average_login = len(data[['user_id', 'date']].drop_duplicates()) / data['user_id'].nunique()

headers = ['加购次数', '移出购物车次数', '购买次数', '平均使用天数']
behaviour_sum = [[cart_n, rmc_n, purchase_n, f"{month_average_login:.2f}"]]
behaviour_sum_table = (
    Table()
    .add(headers, behaviour_sum)
)
behaviour_sum_table.render_notebook()

加购次数,移出购物车次数,购买次数,平均使用天数
3171756,1652311,727120,1.61


#### 交易指标

指标口径：
- 成交额：购买物品的价格总数
- 客单价：成交额除以购买客户数
- 客户数：有购买记录的用户数
- 购买率：所有用户中，有购买记录的用户占比
- 复购客户数：有购买记录的用户中，有两次以上购买记录的用户
- 复购率：有购买记录的用户中，有两次以上购买记录的用户占比
- 平均购买次数：购买用户的平均购买次数

In [4]:
turnovers = data['turnovers'].sum()
user_gb = data[['user_id', 'view', 'cart', 'remove_from_cart', 'purchase', 'turnovers']].groupby('user_id').sum()
buyer_gb = user_gb[user_gb['purchase'] > 0]
customer_unit_price = turnovers / buyer_gb.shape[0]
purchase_rate = len(buyer_gb) / len(user_gb)
repurchase_n = len(buyer_gb[buyer_gb['purchase'] > 1])
repurchase_rate = repurchase_n / len(buyer_gb)

headers = ['成交额', '客单价', '客户数', '购买率', '复购客户数', '复购率', '客户平均购买次数']
transaction = [[
    f"{turnovers:.2f}",
    f"{customer_unit_price:.2f}",
    f"{buyer_gb.shape[0]}",
    f"{purchase_rate:.2%}",
    f"{repurchase_n}",
    f"{repurchase_rate:.2%}",
    f"{buyer_gb['purchase'].mean():.0f}"
]]
transaction_table = (
    Table()
    .add(headers, transaction)
)
transaction_table.render_notebook()

成交额,客单价,客户数,购买率,复购客户数,复购率,客户平均购买次数
3585051.20,50.03,71654,7.39%,63915,89.20%,10


**结论：整体购买率仅 7.39%，新用户的运营有改善空间；复购率高达 89.20%，老客户的粘性很高。**

### 整体行为转化

In [5]:
def draw_behaviour_funnel(data, behaviour_lst, overall=True, is_show=True, title='Funnel'):
    behaviour_df = pd.DataFrame(data[behaviour_lst].sum(), index=behaviour_lst, columns=['n'])
    if overall:
        behaviour_df['转化率'] = behaviour_df['n'] / behaviour_df.iloc[0, 0]
    else:
        behaviour_df['转化率'] = behaviour_df['n'].shift(-1) / behaviour_df['n']
        behaviour_df['转化率'] = behaviour_df['转化率'].shift(1)
        behaviour_df.iloc[0, 1] = 1
    behaviour_df['转化率'] = behaviour_df['转化率'].apply(lambda x: format(x, ".2%"))
    behaviour_df['转化环节'] = behaviour_df.index + ': ' + behaviour_df['转化率']

    behaviour_funnel = (
        Funnel()
        .add(
            '',
            behaviour_df[['转化环节', 'n']].values,
        )
        .set_global_opts(
            title_opts=opts.TitleOpts(
                title=title,
                subtitle='转化率计算方式：每个环节以第一个环节作为计算基准' if overall else '计算方式：每个环节以前一个环节作为计算基准',
            ),
            legend_opts=opts.LegendOpts(
                pos_left='right',
                pos_top='bottom',
                orient='vertical'
            ),
        )
        .set_series_opts(
            label_opts=opts.LabelOpts(
                position='inner',
                formatter='{b}',
            )
        )
    )
    return behaviour_funnel.render_notebook() if is_show else behaviour_funnel

In [6]:
# title = '浏览-加购-购买整体转化'
# draw_behaviour_funnel(data, ['view', 'cart', 'purchase'], title=title)

In [7]:
# title = '浏览-加购-移出购物车整体转化'
# draw_behaviour_funnel(data, ['view', 'cart', 'remove_from_cart'], title=title)

In [8]:
convert_df = data[['view', 'cart', 'purchase', 'remove_from_cart']].sum()
convert_df = np.round(convert_df / convert_df[0] * 100, 2)

colors = ['#D53A35', '#2F4554', '#61A0A8', '#D48265', '#749F83', '#CA8622', '#BDA29A', '#6E7074']

behaviour_nodes = [
    {"name": "view"},
    {"name": "cart"},
    {"name": "remove_from_cart"},
    {"name": "purchase"},
    {"name": "miss"},
]

behaviour_links = [
    {"source": "view", "target": "cart", "value": convert_df['cart']},
    {"source": "view", "target": "miss", "value": np.round(convert_df['view'] - convert_df['cart'], 2)},
    {"source": "cart", "target": "purchase", "value": convert_df['purchase']},
    {"source": "cart", "target": "remove_from_cart", "value": convert_df['remove_from_cart']},
    {"source": "cart", "target": "miss", "value": np.round(convert_df['cart'] - convert_df['purchase'] - convert_df['remove_from_cart'], 2)},
]

behaviour_sankey = (
    Sankey()
    .set_colors(colors=colors)
    .set_global_opts(
        title_opts=opts.TitleOpts(
            title="整体行为转化流向",
            subtitle="注：转化率以【浏览量】为计算基准。",
        ),
        tooltip_opts=opts.TooltipOpts(trigger="item", trigger_on="mousemove"),
        
    )
    .add(
        "", 
        behaviour_nodes, 
        behaviour_links,
        orient="vertical",
        pos_top="10%",
        pos_left="10%",
        pos_right="10%",
        focus_node_adjacency="allEdges",
        linestyle_opt=opts.LineStyleOpts(opacity=0.4, curve=0.4, color="target"),
    )
)
behaviour_sankey.render_notebook()

结论：
- “浏览-加购”的转化率为 60.34%，反映出用户浏览后多半会使用加购功能。
- “加购-移出购物车”的整体转换率为 31.43%，是“加购-购买”的整体转化率 13.83% 的两倍不止，猜测用户加购的目的不单是计划购买，还可能是商品收藏或价格观察，所以加购后又予以移除。

建议：
- 推出收藏功能，既完善用户使用的功能性，又便于我们分析用户的意图。
- 适时调整商品价格，促进用户购买。例如：发放优惠券促销、清仓打折。
- 当加购后的商品发生信息变更或久未操作，及时提醒用户。例如：客户端-推送、web 端-消息通知。

## 时间维度分析

In [9]:
def show_time_trend(uv_df, behaviour_df, bar_title='UV', bar_subtitle='-', line_title='Behaviour', line_subtitle='-', datazoom_opts_show=False, is_show=True):
    """Show the data trend of specific timescales."""
    uv_bar = (
        Bar()
        .set_global_opts(
            title_opts=opts.TitleOpts(
                title=bar_title,
                subtitle=bar_subtitle,
            ),
            tooltip_opts=opts.TooltipOpts(
                is_show=True, trigger="axis", axis_pointer_type="cross"
            ),
            xaxis_opts=opts.AxisOpts(
                type_="category",
                axispointer_opts=opts.AxisPointerOpts(is_show=True, type_="shadow"),
            ),
        )
        .add_xaxis(uv_df.index.astype('str').tolist())
    )
    for c in uv_df.columns:
        uv_bar.add_yaxis(c, uv_df[c].tolist())
        uv_bar.set_series_opts(
            label_opts=opts.LabelOpts(is_show=False),
            markpoint_opts=opts.MarkPointOpts(
                data=[
                    opts.MarkPointItem(type_="max", name="最大值"),
                    opts.MarkPointItem(type_="min", name="最小值"),
                ]
            ),
            markline_opts=opts.MarkLineOpts(
                data=[
                    opts.MarkLineItem(name=f"{c}均值", type_="average")
                ],
                symbol='none',
                label_opts=opts.LabelOpts(is_show=False),
                linestyle_opts=opts.LineStyleOpts(
                    opacity=0.5,
                    type_='dashed'
                ),
            ),
        )
    if datazoom_opts_show:
        uv_bar.set_global_opts(datazoom_opts=opts.DataZoomOpts(is_show=datazoom_opts_show))
    
    behaviour_line = (
        Line()
        .set_global_opts(
            title_opts=opts.TitleOpts(
                title=line_title,
                subtitle=line_subtitle,
            ),
            tooltip_opts=opts.TooltipOpts(
                is_show=True, trigger="axis", axis_pointer_type="cross"
            ),
        )
        .add_xaxis(behaviour_df.index.astype(str).tolist())
    )
    for b in behaviour_df.columns:
        behaviour_line.add_yaxis(b, behaviour_df[b].tolist())
        behaviour_line.set_series_opts(
            label_opts=opts.LabelOpts(is_show=False),
            linestyle_opts=opts.LineStyleOpts(width=2),
            markpoint_opts=opts.MarkPointOpts(
                data=[
                    opts.MarkPointItem(type_="max", name="最大值"),
                    opts.MarkPointItem(type_="min", name="最小值"),
                ]
            ),
            markline_opts=opts.MarkLineOpts(
                data=[
                    opts.MarkLineItem(name=f"{b}均值", type_="average")
                ],
                symbol='none',
                label_opts=opts.LabelOpts(is_show=False),
                linestyle_opts=opts.LineStyleOpts(
                    opacity=0.5,
                    type_='dashed'
                ),
            ),
        )

    page = (
        Page()
            .add(uv_bar, behaviour_line)
    )
    return page.render_notebook() if is_show else page

In [10]:
def double_y(data, left, right, title='Double y-axis', left_kind=Line(), right_kind=Line(), is_show=True):
    """双 y 轴"""
#     kind_mapper = {
#         'Line': Line(),
#         'Bar': Bar(),
#     }
#     left_chart = kind_mapper[left_kind] 
#     right_chart = kind_mapper[right_kind]
    
    left_chart = (
        Line()
        .add_xaxis(data.index.astype('str').tolist())
        .set_global_opts(
            title_opts=opts.TitleOpts(
                title=title,
            ),
            tooltip_opts=opts.TooltipOpts(
                is_show=True, trigger="axis", axis_pointer_type="cross"
            ),
            xaxis_opts=opts.AxisOpts(
                name=f"{data.index.name}",
                name_location='middle',
                name_gap=30,
                type_="category",
                axispointer_opts=opts.AxisPointerOpts(is_show=True, type_="shadow"),
            ),
            yaxis_opts=opts.AxisOpts(
                name=f"{'/'.join(left)}",
            ),
        )
        .extend_axis(
            yaxis=opts.AxisOpts(
                name=f"{'/'.join(right)}",
                axislabel_opts=opts.LabelOpts(formatter="{value}"),
            )
        )
    )

    for l in left:
        left_chart.add_yaxis(
            l,
            data[l].tolist(),
            label_opts=opts.LabelOpts(is_show=False),
            linestyle_opts=opts.LineStyleOpts(width=2),
        )

    right_chart = (
        Line()
        .add_xaxis(data.index.astype('str').tolist())
    )

    for r in right:
        right_chart.add_yaxis(
            r,
            data[r].tolist(),
            yaxis_index=1,
            linestyle_opts=opts.LineStyleOpts(curve=360, type_='dotted'),
            label_opts=opts.LabelOpts(is_show=False),
        )

    left_chart.overlap(right_chart)
    return left_chart.render_notebook() if is_show else left_chart

### 按月分析

In [11]:
time_type = 'month'
uv_df = pd.DataFrame(data.groupby(time_type)['user_id'].apply(lambda x: x.nunique())).rename(columns={'user_id': 'uv'})
behaviour_df = data.groupby(time_type)[['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers']].sum()
bar_title, bar_subtitle = f"每月 PV/UV", '注：汇总方式 - 按月求和'
line_title, line_subtitle = f"每月行为指标", '注：汇总方式 - 按月求和'
uv_df['pv'] = behaviour_df['view']
uv_df = uv_df[['pv', 'uv']]
show_time_trend(uv_df, behaviour_df, bar_title, bar_subtitle, line_title, line_subtitle)

现象：第四季度中，每月各项指标的涨跌趋势一致，整体上，11 月份表现最好，12 月份表现最差。

初步推测：11 月份活动力度较大，刺激用户活跃和消费，也因此 12 月份数据较为疲软。

### 按周分析

In [12]:
week_day_n = data.groupby('week')['date'].apply(lambda x: x.nunique())

time_type = 'week'
uv_df = pd.DataFrame(data.groupby(time_type)['user_id'].apply(lambda x: x.nunique())).rename(columns={'user_id': 'uv'})
behaviour_df = data.groupby(time_type)[['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers']].sum()
uv_df.drop(index=1, inplace=True)
behaviour_df.drop(index=1, inplace=True)
bar_title, bar_subtitle = f"每周的日均 PV/UV", '注：汇总方式 - 按周求和再除以当周天数；该月第 44 周仅 3 天，第 48 周仅 6 天。'
line_title, line_subtitle = f"每周的日均行为指标汇总", '注：汇总方式 - 按周求和再除以当周天数；该月第 44 周仅 3 天，第 48 周仅 6 天。'
uv_df['pv'] = behaviour_df['view']

uv_df = uv_df[['pv', 'uv']]
for c in uv_df.columns:
    uv_df[c] = round(uv_df[c] / week_day_n)
for c in behaviour_df.columns:
    behaviour_df[c] = round(behaviour_df[c] / week_day_n)
show_time_trend(uv_df, behaviour_df, bar_title, bar_subtitle, line_title, line_subtitle)

现象：
1. 属于 11 月份的 47 周、48 周的数据明显高于平均值和其余周，是 11 月份表现最好的原因。
2. 属于 12 月份的 51 周、52 周的数据均低于其它周且呈明显的连续下降，12 月份表现最差。
3. 除去以上时间，每周指标均围绕各周均值小幅波动。

结论：
1. 40 ~ 46 周的数据比较稳定，更具日常代表性。
2. 47 周、48 周的高峰期和 51 周、52 周的连续衰退使得 [11 月份表现最好，12 月份表现最差](#按月分析)。

### 按星期分析

In [13]:
weekday_n = data.groupby('weekday')['week'].apply(lambda x: x.nunique())

time_type = 'weekday'
uv_df = pd.DataFrame(data.groupby(time_type)['user_id'].apply(lambda x: x.nunique())).rename(columns={'user_id': 'uv'})
behaviour_df = data.groupby(time_type)[['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers']].sum()
bar_title, bar_subtitle = f"日均星期 PV/UV", '注：汇总方式 - 按星期求和再除以该星期次数；该月有两个星期四和星期五。'
line_title, line_subtitle = f"日均星期行为指标", '注：汇总方式 - 按星期求和再除以该星期次数；该月有两个星期和四星期五。'
uv_df['pv'] = behaviour_df['view']
uv_df = uv_df[['pv', 'uv']]
for c in uv_df.columns:
    uv_df[c] = round(uv_df[c] / weekday_n)
for c in behaviour_df.columns:
    behaviour_df[c] = round(behaviour_df[c] / weekday_n)

In [14]:
time_type = 'weekday'
weekday_sort = ['星期一', '星期二', '星期三', '星期四', '星期五', '星期六', '星期日', ]
weekday_n = data.groupby('weekday')['week'].apply(lambda x: x.nunique())
uv_df = pd.DataFrame(data.groupby(time_type)['user_id'].apply(lambda x: x.nunique())).rename(columns={'user_id': 'uv'}).reindex(weekday_sort)
behaviour_df = data.groupby(time_type)[['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers']].sum().reindex(weekday_sort)
bar_title, bar_subtitle = f"日均星期 PV/UV", '注：汇总方式 - 按星期求和再除以该星期次数；该月有两个星期四和星期五。'
line_title, line_subtitle = f"日均星期行为指标", '注：汇总方式 - 按星期求和再除以该星期次数；该月有两个星期和四星期五。'
uv_df['pv'] = behaviour_df['view']
uv_df = uv_df[['pv', 'uv']]
for c in uv_df.columns:
    uv_df[c] = round(uv_df[c] / weekday_n)
for c in behaviour_df.columns:
    behaviour_df[c] = round(behaviour_df[c] / weekday_n)
show_time_trend(uv_df, behaviour_df, bar_title, bar_subtitle, line_title, line_subtitle)

现象：整体上，各星期平均数据较为平缓，围均值小幅波动。值得注意的是周四、周五成交额明显高于均值，而周六、周日的各项数据都均低于均值，其中周六的各项数据是最低点。

初步推测：活动多在周四、周五推出，因此周四、周五平均成交额较高，也导致周六各项平均数据短暂疲软。

### 按日期分析

In [15]:
time_type = 'date'
uv_df = pd.DataFrame(data.groupby(time_type)['user_id'].apply(lambda x: x.nunique())).rename(columns={'user_id': 'uv'})
behaviour_df = data.groupby(time_type)[['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers']].sum()
bar_title, bar_subtitle = f"每日 PV/UV", '注：汇总方式 - 按日期求和'
line_title, line_subtitle = f"每日行为指标", '注：汇总方式 - 按日期求和'
uv_df['pv'] = behaviour_df['view']
uv_df = uv_df[['pv', 'uv']]
show_time_trend(uv_df, behaviour_df, bar_title, bar_subtitle, line_title, line_subtitle)

现象：
1. 整体上，每日各项指标数据大约以**七天**为一个周期呈现周期性波动。
2. 本季度存在 3 个成交额的高峰期，根据峰值由高到低依次为： 2019-11-21 ~ 2019-11-24、2019-11-07 ~ 2019-11-08、 2019-11-07 ~ 2019-11-08。
3. 2019-11-16 过后各项数据呈明显周期性下滑趋势。

In [16]:
import datetime
behaviour_type = 'purchase'
calendar_begin = np.min(uv_df.index)
calendar_end = np.max(uv_df.index)
calendar_values = behaviour_df[behaviour_type].values.tolist()
calendar_data = [[str(calendar_begin + datetime.timedelta(days=i)), calendar_values[i]]
        for i in range((calendar_end - calendar_begin).days + 1)]
        
behaviour_calendar = (
    Calendar()
    .set_global_opts(
        title_opts=opts.TitleOpts(title=f'每日 {behaviour_type} 数据'),
        tooltip_opts=opts.TooltipOpts(
            is_show=True,
            formatter='{c}'
        ),
        visualmap_opts=opts.VisualMapOpts(
            min_=min(calendar_values),
            max_=max(calendar_values),
            is_calculable=True,
            orient='horizontal',
            pos_top='bottom',
            pos_left='right',
            precision=0,
        ),
    )
    .add(
        behaviour_type,
        calendar_data,
        calendar_opts=opts.CalendarOpts(
            range_=[calendar_begin, calendar_end],
            orient='vertical',
            pos_bottom='20%',
        ),
    )
)
behaviour_calendar.render_notebook()

现象：
1. 2019-11 月份活动频繁：
   - 2019-11-21 ~ 2019-11-24 期间数据明显优于其它时间段，其中 2019-11-22 当天数据表现最好，经查证，当天是黑色星期五，而感恩节（2019-11-21）后的几天是购物高峰期。
   > 注：黑五一般指黑色星期五。 美国圣诞节大采购一般是从感恩节之后开始的。感恩节是每年11月的第四个星期四，因此它的第二天，也就是美国人大采购的第一天。在这一天，美国的商场都会推出大量的打折和优惠活动，以在年底进行最后一次大规模的促销。
   - 除去黑五活动期，其余两次活动期 2019-11-07 ~ 2019-11-08、 2019-11-07 ~ 2019-11-08 均发生在在周四~周六之间。
2. 除去活动期，工作日的数据表现优于周末。
3. 2019-12 月份下旬的数据表现呈明显逐日递减，2019-11-06 过后下降尤其明显，推测是圣诞节刚结束，购物需求降低。

In [17]:
user_daily = pd.DataFrame(data.groupby('date')['user_id'].apply(lambda x: x.nunique()))
turnovers_day = pd.DataFrame(data.groupby('date')['turnovers'].sum())
user_daily = pd.concat([user_daily, turnovers_day], axis=1)
user_daily.columns = ['DAU', '日成交额']
user_daily['日客单价'] = user_daily['日成交额'] / user_daily['DAU']
user_daily.index.name = '日期'

# double_y(user_daily, left=['日客单价'], right=['DAU', '日成交额'], title='日客单价和 DAU 变化')
double_y(user_daily, left=['日客单价'], right=['DAU'], title=' DAU 和每日客单价变化')

现象：圣诞节过后，DAU 和日客单价都明显降低，前者可以看出用户购物需求减少，后者可以看出客户消费能力疲软。

**结论：**
- **黑五是用户季度活跃和成交的最高峰，圣诞节过后用户活跃和成交持续走低。**
- **活动期间，除了流量、销量，用户客单价也会明显提高。**

In [18]:
# dut_data = list()
# dut = data.groupby(['date', 'user_id'])[['turnovers']].sum()
# dut = dut[dut['turnovers'] > 0].reset_index().drop(columns=['user_id'])
# date_list = dut['date'].astype(str).unique().tolist()
# for i in dut['date'].unique():
#     dut_data.append(list(dut[dut['date'] == i]['turnovers']))

# dut_kline = (
#     Kline()
#     .set_global_opts(
#         title_opts=opts.TitleOpts(title='用户日成交额'),
#         xaxis_opts=opts.AxisOpts(name='日期'),
#         yaxis_opts=opts.AxisOpts(name='用户日成交额'),
#         datazoom_opts=opts.DataZoomOpts(
#             is_show=True,
#             range_start=0,
#             range_end=100,
#         ),
#     )
#     .add_xaxis(date_list)
#     .add_yaxis('', dut_data)
# )
# dut_kline.render_notebook()

### 按时段分析

In [19]:
time_type = 'hour'
uv_df = pd.DataFrame(data.groupby(time_type)['user_id'].apply(lambda x: x.nunique())).rename(columns={'user_id': 'uv'})
behaviour_df = data.groupby(time_type)[['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers']].sum()
bar_title, bar_subtitle = f"各时段 PV/UV", '注：汇总方式 - 按时段求和'
line_title, line_subtitle = f"各时段行为指标", '注：汇总方式 - 按时段求和'
uv_df['pv'] = behaviour_df['view']
uv_df = uv_df[['pv', 'uv']]
show_time_trend(uv_df, behaviour_df, bar_title, bar_subtitle, line_title, line_subtitle)

In [20]:
behaviour_df['客单价'] = np.round(behaviour_df['turnovers'] / uv_df['uv'], 4)
behaviour_df['浏览-购买转化率'] = np.round(behaviour_df['purchase'] / behaviour_df['view'], 4)
behaviour_df['浏览-加购转化率'] = np.round(behaviour_df['cart'] / behaviour_df['view'], 4)
behaviour_df['加购-移除转化率'] = np.round(behaviour_df['remove_from_cart'] / behaviour_df['cart'], 4)
behaviour_df.index.name = '时段'

double_y(
    behaviour_df,
    left=['客单价'],
    right=['浏览-购买转化率', '浏览-加购转化率', '加购-移除转化率'],
    title='时段客单价与转化率',
)

现象：
1. 11 时 为日高峰，19 时为夜高峰
2. 夜里 20 时过后数据持续下降，至夜里 0、1 时低至当日最低；
3. 2 时起逐步回升，于 11 时达到日高峰；
4. 11 过后至 15 时小幅回落，16 时起回升，至 19 时达到夜高峰。
5. 高峰期的客单价略高于其它时段。
6. 除去深夜（0 时至 6 时），各时段环节转化率变化平缓，波动不明显。

**结论：11 时和 19 时左右是每日用户活跃与成交的高峰期，客单价高于其它时段，但各环节转化率没有提升。**

建议：
- 注重日高峰和夜高峰（尤其是夜高峰）的运营：
    - 在该时段进行活动和商品推送，如：新品上线、商品限时折扣。
    - 在该时段限时价格优惠，促进用户购买，如：如促销或库存类商品优惠券。

## 用户分析

### 用户行为统计信息

用户汇总后的行为统计信息：

In [21]:
user_gb.describe()

,view,cart,remove_from_cart,purchase,turnovers
count,969369.000000,969369.000000,969369.000000,969369.000000,969369.000000
mean,5.422704,3.271980,1.704522,0.750096,3.698335
std,29.082329,16.531505,13.350551,4.626981,22.866726
min,0.000000,0.000000,0.000000,0.000000,0.000000
25%,1.000000,0.000000,0.000000,0.000000,0.000000
50%,1.000000,0.000000,0.000000,0.000000,0.000000
75%,3.000000,1.000000,0.000000,0.000000,0.000000
max,15518.000000,1885.000000,2448.000000,417.000000,2601.090000


可见，不同用户的行为指标差异可能极大。查看未产生购买行为用户与客户的中位值：

根据均值-方差方式（超出均值 3 个标准差的区间视为异常值）去除异常值后的统计信息：

In [22]:
user_gb[user_gb['purchase'] > 0].describe()

,view,cart,remove_from_cart,purchase,turnovers
count,71654.000000,71654.000000,71654.000000,71654.000000,71654.000000
mean,31.074176,25.277361,14.876434,10.147654,50.032813
std,70.926584,45.825653,40.577388,13.938066,68.961423
min,0.000000,0.000000,0.000000,1.000000,0.060000
25%,4.000000,5.000000,1.000000,3.000000,15.490000
50%,12.000000,11.000000,4.000000,6.000000,30.560000
75%,31.000000,27.000000,14.000000,12.000000,55.830000
max,5245.000000,1649.000000,2448.000000,417.000000,2601.090000


In [23]:
user_gb[user_gb['purchase'] == 0].describe()

,view,cart,remove_from_cart,purchase,turnovers
count,897715.000000,897715.000000,897715.000000,897715.000000,897715.000000
mean,3.375249,1.515550,0.653164,0.000000,0.000000
std,21.331939,9.260288,6.788894,0.000000,0.000000
min,0.000000,0.000000,0.000000,0.000000,0.000000
25%,1.000000,0.000000,0.000000,0.000000,0.000000
50%,1.000000,0.000000,0.000000,0.000000,0.000000
75%,3.000000,0.000000,0.000000,0.000000,0.000000
max,15518.000000,1885.000000,1650.000000,0.000000,0.000000


In [24]:
def u_3sigma(data):
    data_outliars = pd.DataFrame()
    for f in data.columns:
        u, sigma = data[f].mean(), data[f].std()
        lower, upper = u - 3 * sigma, u + 3 * sigma
        data_outliars[f] = data[f].apply(lambda x: False if x >= lower and x <= upper else True)
    return data_outliars


user_outliars = u_3sigma(user_gb)
user_clear = user_gb[~user_outliars.any(axis=1)]
user_clear
user_clear.describe()

,view,cart,remove_from_cart,purchase,turnovers
count,943274.000000,943274.000000,943274.000000,943274.000000,943274.000000
mean,3.551597,1.622310,0.619613,0.260682,1.362734
std,7.303044,4.859485,2.956417,1.343775,6.828103
min,0.000000,0.000000,0.000000,0.000000,0.000000
25%,1.000000,0.000000,0.000000,0.000000,0.000000
50%,1.000000,0.000000,0.000000,0.000000,0.000000
75%,3.000000,1.000000,0.000000,0.000000,0.000000
max,92.000000,52.000000,41.000000,14.000000,72.290000


### 用户活跃情况

In [48]:
user_day_n = data.groupby('user_id')[['date']].agg(['min', 'max'])
user_day_n.columns = ['季度首日登录', '季度最后一日登录']
user_day_n = pd.DataFrame(
    [user_day_n['季度首日登录'].value_counts().sort_index(),
    user_day_n['季度最后一日登录'].value_counts().sort_index()]
).T

day_new_line = (
    Line()
    .set_global_opts(
        title_opts=opts.TitleOpts(
            title='当季首日登录/最后一日登录用户数'
        ),
        tooltip_opts=opts.TooltipOpts(
            is_show=True, trigger="axis", axis_pointer_type="cross"
        ),
        xaxis_opts=opts.AxisOpts(
            type_="category",
            axispointer_opts=opts.AxisPointerOpts(is_show=True, type_="shadow"),
        ),
    )
    .add_xaxis(user_day_n.index.astype('str').tolist())
    .add_yaxis('季度首日登录', user_day_n['季度首日登录'].tolist())
    .add_yaxis('季度最后一日登录', user_day_n['季度最后一日登录'].tolist())
    .set_series_opts(label_opts=opts.LabelOpts(is_show=False))
)
day_new_line.render_notebook()

现象：活动期间，有大量用户首次登录，但也有大量用户活动期过后不再登录。

结论：部分用户仅在活动期活跃。

建议：合理安排活动期，同时做好拉新与召回。
- 拉新：
    1. 以老带新，邀请双方都有优惠
    2. 通过社交媒体、广告增大曝光量，吸引新用户
- 召回：
    1. 针对流失用户（长周期内未使用）：在重大活动或改版的时候，以邮箱、电话、推广的方式召回
    2. 针对沉默用户（长周期内使用过，短周期内未使用）：通过内容推荐、定期活动推送召回

In [26]:
user_new_day = data.groupby('user_id')[['date']].min()
user_new_day.columns = ['first_day']
user_day = pd.merge(user_new_day, data[['user_id', 'date']], on='user_id', how='left')
user_day['nday'] = user_day['date'] - user_day['first_day']
user_day = user_day[user_day['nday'].dt.days < 31]  # 只取 30 天留存
day_retention_n = user_day[['first_day', 'nday', 'user_id']].groupby(['first_day', 'nday'], as_index=False).count()
# day_retention_n = day_retention_n.reset_index()

retention_table = day_retention_n.pivot_table(index=['first_day'], columns=['nday'])
retention_table.columns = [i for i in range(retention_table.shape[1])]
first_day_n = retention_table[0]
del retention_table[0]
for col in retention_table.columns:
    retention_table[col] = np.round(retention_table[col] / first_day_n, 4)
    
# rentention_rate = retention_table[[1, 3, 7]]
# rate_columns = [f'{col} 日留存率' for col in rentention_rate.columns]
# rentention_rate.columns = rate_columns
# rentention_rate['turnovers'] = data.groupby('day')['turnovers'].sum()
# double_y(rentention_rate, left=rate_columns, right=['turnovers'], title='用户留存')

retention_table = retention_table.iloc[::-1].T
retention_values = [[i, j, val] for i, row in enumerate(retention_table.values) for j, val in enumerate(row)]
max_retention = np.max(retention_table)

retention_heatmap = (
    HeatMap()
    .set_global_opts(
        title_opts=opts.TitleOpts(title='每日用户留存（30日）'),
        xaxis_opts=opts.AxisOpts(
            name='第 N 日',
            position='top'
        ),
        yaxis_opts=opts.AxisOpts(
            name='日期',
            name_location='start',
            axislabel_opts=opts.LabelOpts(font_size=9),
        ),
        visualmap_opts=opts.VisualMapOpts(
            max_=np.round(np.max(retention_table.iloc[0]) + 0.1, 1),
            is_calculable=True,
            pos_top='middle',
            pos_left='right',
            precision=2,
        ),
    )
    .add_xaxis(retention_table.index.astype(str).tolist())
    .add_yaxis("第 N 日留存率", retention_table.columns.astype(str).tolist(), retention_values)
)
retention_heatmap.render_notebook()

现象：次日留存在 0.1~0.2 之间，第一日往后的留存基本都小于 0.1，留存率整体较低且变化不明显，在活动期间有微小提升。

结论：对于电商，大多数用户在有购物需求或活动期才会使用，用户粘性低于具有日常高频的使用需求的社交/游戏类产品，因而难有稳定且高的留存率。

In [27]:
# 访问次数
user_view_cnt = data[['user_id', 'view', 'turnovers']].groupby('user_id').sum()
# 访问天数
user_login_days = data[['user_id', 'date']].drop_duplicates().groupby('user_id').count()
user_login_days.columns = ['login_days']
user_stickiness = pd.concat([user_view_cnt, user_login_days], axis=1)
user_stickiness['average_daily_view'] = user_stickiness['view'] / user_stickiness['login_days']
# user_stickiness
login_days_df = user_stickiness.groupby('login_days')[['turnovers']].mean()
login_days_df.index.name = '累计登陆天数'
login_days_df.columns = ['平均成交额']
login_days_df['用户比例'] = user_stickiness['login_days'].value_counts(normalize=True).sort_index()
login_days_df['平均日均访问量'] = user_stickiness.groupby('login_days')[['average_daily_view']].mean()
double_y(login_days_df, left=['用户比例'], right=['平均成交额', '平均日均访问量'], title='用户累计登陆天数分布')

现象：
- 随着登陆天数的增大，相应的用户数量迅速降低：当月 78% 的用户仅登录 1 天，11% 的用户仅登录 2 天，仅有约 5 % 的用户登陆天数大于 3 天
- 整体上，登陆天数越大，用户的日均访问量与成交额越高

**结论：提高用户日常粘性，以期提高单个用户的 LTV。**

建议：针对登陆天数少的用户采取措施进行优化。除去电商本就不如即时通讯类具有日常高频的使用需求，导致用户登陆天数少的原因或许有用户使用体验不佳、对内容不感冒等。优化方向可能有：
1. 推出日常任务体系（如：签到、浏览可得积分并换取优惠券）
2. 针对提优化界面和操作，提高用户的好感
3. 提高推荐系统精准度，适时推送内容

### 用户行为路径分布

In [28]:
up_path = data.groupby(['user_id', 'product_id'])['event_type'].apply(lambda x: ' -> '.join(x))
up_path = pd.DataFrame(up_path)
# up_path
path_distribution = up_path[['event_type']].value_counts(normalize=True)
# path_distribution

path_pie = (
    Pie()
    .add(
        '',
        [list(z) for z in zip(path_distribution[path_distribution > 0.01].index.tolist(), path_distribution[path_distribution > 0.01].values.tolist())],
        radius=[100, 180],
    )
    .set_global_opts(
        title_opts=opts.TitleOpts(
            title='用户-物品行为路径占比',
            subtitle='注：仅展示占比 ≥1% 的行为路径',
        ),
        legend_opts=opts.LegendOpts(is_show=False)
    )
    .set_series_opts(label_opts=opts.LabelOpts(formatter='{b}: {d}%'))
)
path_pie.render_notebook()

现象：（只关注百分比大于 1% 的行为路径）
- 用户-物品交互不一定以浏览（view）为起点，“仅加购”和“加购后移除”分别占比 11.86% 和 11.52%，仅次于“一次浏览”。
- “一次浏览”（51.36%）、“二次浏览”（7.13%）、“三次浏览”（1.66%）占比均大于 1%，在所有路径中分列 1、4、8。
- 用户加购后续行为：没有操作（11.86%） > 移除（11.52%） > 购买（6.53%）。
- 包含购买（purchase）行为的占比最高的的路径是“加购后购买（cart -> purchase）”，占比 6.53% 。

**结论：**
- **用户浏览与重复浏览的转化率低，商品的首次曝光和重复曝光策略应该进行调整。**
- **加购后购买的转化率明显低于没有后续操作和取消购物车，推测用户“加购”不一定代表较高的购买欲望，也可能是收藏商品或价格观察。**

建议：
- 提高推荐系统精准度：使首次曝光的推荐商品更符合用户需求；对曝光后用户可能不感兴趣的商品避免重复曝光。
- 推出收藏功能，既完善用户使用的功能性，又便于我们分析用户的意图。
- 当加购后的商品发生信息变更（活动、降价）或久未操作，及时提醒用户。例如：通过手机、邮箱等联系方式推送。

## 客户分析

### 用户购买情况

In [29]:
udp = data.groupby(by=['user_id', 'date'], as_index=False)[['purchase']].sum()
# udp
ufpd = data[data['purchase'] == 1].groupby(by=['user_id'], as_index=False)[['date']].min()
# ufpd
udp = pd.merge(left=udp, right=ufpd, how='left', on=['user_id'], suffixes=('', '_first_purchase'))
udp['no_purchase'] = (udp['date_first_purchase'].isnull()).astype(int)
udp['new_purchase'] = (udp['date_first_purchase'] == udp['date']).astype(int)
udp['repurchase'] = (udp['date_first_purchase'] < udp['date']).astype(int)
# udp
dus = udp.groupby(by=['date'], as_index=False)[['no_purchase', 'new_purchase', 'repurchase']].sum()
# dus

In [30]:
dus_cumsum_line = (
    Line()
    .set_global_opts(
        title_opts=opts.TitleOpts(title="每日用户购买情况"),
        tooltip_opts=opts.TooltipOpts(trigger="axis", axis_pointer_type="cross"),
        xaxis_opts=opts.AxisOpts(
            name='日期',
            type_="category",
            boundary_gap=False,
        ),
        yaxis_opts=opts.AxisOpts(
            name='用户数',
            type_="value",
            axistick_opts=opts.AxisTickOpts(is_show=True),
            splitline_opts=opts.SplitLineOpts(is_show=True),
        ),
    )
    .add_xaxis(dus['date'].tolist())
    .add_yaxis(
        series_name='当日无购买用户',
        stack='总量',
        y_axis=dus['no_purchase'],
        areastyle_opts=opts.AreaStyleOpts(opacity=0.6),
        label_opts=opts.LabelOpts(is_show=False),
        color='#AAAAAA',
    )
    .add_yaxis(
        series_name='首日购买用户',
        stack='总量',
        y_axis=dus['new_purchase'],
        areastyle_opts=opts.AreaStyleOpts(opacity=0.6),
        label_opts=opts.LabelOpts(is_show=False),
    )
    .add_yaxis(
        series_name='非首日购买用户',
        stack='总量',
        y_axis=dus['repurchase'],
        areastyle_opts=opts.AreaStyleOpts(opacity=0.6),
        label_opts=opts.LabelOpts(is_show=False),
    )
)
dus_cumsum_line.render_notebook()

现象：
- 每日属于重复购买的用户，随着时间推移越来越多。
- 每日属于无购买和初次购买的用户随时间推移无明显增减趋势。

In [31]:
upd = pd.merge(left=data[data['purchase'] == 1][['user_id', 'date']], right=ufpd, how='left', on=['user_id'], suffixes=('', '_first'))
upd['first_date_purchase'] = (upd['date'] == upd['date_first']).astype(int)
upd['repurchase'] = 1 - upd['first_date_purchase']
upd = upd.groupby('date', as_index=False)[['first_date_purchase', 'repurchase']].sum()
# upd['purchase'] = upd['first_date_purchase'] + upd['repurchase']
# first_date_purchase_lst = list()
# repurchase_lst = list()
# for i in upd.index:
#     first_date_purchase_lst.append({"value": upd.loc[i]['first_date_purchase'], "percent": upd.loc[i]['first_date_purchase'] / (upd.loc[i]['repurchase'] + upd.loc[i]['first_date_purchase'])})
#     repurchase_lst.append({"value": upd.loc[i]['repurchase'], "percent": upd.loc[i]['repurchase'] / upd.loc[i]['purchase']})
# print(first_date_purchase_lst)
# print(repurchase_lst)

In [32]:
upd_cumsum_bar = (
    Bar()
    .set_global_opts(
        title_opts=opts.TitleOpts(title="每日购买类型"),
        tooltip_opts=opts.TooltipOpts(trigger="axis", axis_pointer_type="cross"),
        yaxis_opts=opts.AxisOpts(
            type_="value",
            axistick_opts=opts.AxisTickOpts(is_show=True),
            splitline_opts=opts.SplitLineOpts(is_show=True),
        ),
        xaxis_opts=opts.AxisOpts(type_="category", boundary_gap=False),
    )
    .add_xaxis(upd['date'].tolist())
    .add_yaxis(
        "首日购买",
        upd['first_date_purchase'].tolist(),
        stack="总量",
        category_gap="50%",
        label_opts=opts.LabelOpts(is_show=False),
    )
    .add_yaxis(
        
        "非首日购买",
        upd['repurchase'].tolist(),
        stack="总量",
        category_gap="50%",
        label_opts=opts.LabelOpts(is_show=False),
    )
)
upd_cumsum_bar.render_notebook()

In [33]:
daily_purchase_line = (
    Line()
    .set_global_opts(
        title_opts=opts.TitleOpts(title="日均客户购买次数"),
        tooltip_opts=opts.TooltipOpts(trigger="axis", axis_pointer_type="cross"),
        xaxis_opts=opts.AxisOpts(
            name='日期',
            type_="category",
            boundary_gap=False,
        ),
        yaxis_opts=opts.AxisOpts(
            name='购买次数',
            type_="value",
            axistick_opts=opts.AxisTickOpts(is_show=True),
            splitline_opts=opts.SplitLineOpts(is_show=True),
        ),
    )
    .add_xaxis(upd['date'].tolist())
    .add_yaxis(
        '首日购买',
        np.floor(upd['first_date_purchase'] / dus['new_purchase']).values.tolist(),
        label_opts=opts.LabelOpts(is_show=False),
    )
    .add_yaxis(
        '非首日购买', 
        np.floor(upd['repurchase'] / dus['repurchase']).values.tolist(),
        label_opts=opts.LabelOpts(is_show=False),
    )
)
daily_purchase_line.render_notebook()

**结论：客户在首日购买的商品数量（平均≥5件）明显多于复购日购买的商品数量（平均≤2件）。**

### 每日客户复购

In [34]:
udd = pd.merge(left=ufpd.rename(columns={'date':'first_date'}), right=data[data['purchase'] == 1][['user_id', 'date']].drop_duplicates(), how='left', on=['user_id'], suffixes=('_first', ''))
udd['nday'] = (udd['date'] - udd['first_date']).dt.days
# repurchase_n = udd[udd['nday'] < 31][['first_date', 'nday', 'user_id']].groupby(['first_date', 'nday'], as_index=False).count()
repurchase_n = udd[['first_date', 'nday', 'user_id']].groupby(['first_date', 'nday'], as_index=False).count()
# repurchase_n
repurchase_table = repurchase_n.pivot_table(index=['first_date'], columns=['nday'])
repurchase_table.columns = [i for i in range(repurchase_table.shape[1])]
first_date_n = repurchase_table[0]
del repurchase_table[0]
for col in repurchase_table.columns:
    repurchase_table[col] = np.round(repurchase_table[col] / first_date_n, 4)

repurchase_table = repurchase_table.iloc[::-1].T
repurchase_values = [[i, j, val] for i, row in enumerate(repurchase_table.values) for j, val in enumerate(row)]
max_repurchase = np.max(repurchase_table)
# repurchase_table

In [35]:
repurchase_heatmap = (
    HeatMap()
    .set_global_opts(
        title_opts=opts.TitleOpts(title='用户每日复购率'),
        xaxis_opts=opts.AxisOpts(
            name='第 N 日',
            position='top'
        ),
        yaxis_opts=opts.AxisOpts(
            name='日期',
            name_location='start',
            axislabel_opts=opts.LabelOpts(font_size=9),
        ),
        visualmap_opts=opts.VisualMapOpts(
            max_=np.round(np.max(repurchase_table.iloc[0]) + 0.05, 1),
            is_calculable=True,
            pos_top='middle',
            pos_left='right',
            precision=2,
        ),
    )
    .add_xaxis(repurchase_table.index.astype(str).tolist())
    .add_yaxis("第 N 日复购率", repurchase_table.columns.astype(str).tolist(), repurchase_values)
)
repurchase_heatmap.render_notebook()

现象：整体日复购率均低于 0.1，活动期稍高于日常。

### 购买次数分布

In [4]:
def draw_pareto(data, tittle='帕累托图', x_name='', is_show=True):
    '''
    Draw pareto.
    '''
    data[f'{data.columns[0]}-累计占比'] = np.round(data.values[:, 0].cumsum() / data.values[:, 0].sum(), 2)
    
    bar = (
        Bar()
        .set_global_opts(
            title_opts=opts.TitleOpts(title=tittle),
            tooltip_opts=opts.TooltipOpts(
                is_show=True, trigger="axis", axis_pointer_type="cross"
            ),
            xaxis_opts=opts.AxisOpts(
                name=x_name,
                name_location='center',
                name_gap=30,
                type_="category",
                axispointer_opts=opts.AxisPointerOpts(is_show=True, type_="shadow"),
            ),
            yaxis_opts=opts.AxisOpts(
                name=data.columns[0],
                type_="value",
                axislabel_opts=opts.LabelOpts(formatter="{value}"),
                axistick_opts=opts.AxisTickOpts(is_show=True),
            ),
            legend_opts=opts.LegendOpts(
                type_="scroll",
            ),
        )
        .extend_axis(
            yaxis=opts.AxisOpts(
                name=data.columns[1],
                type_="value",
                min_=0,
                max_=1.1,
                axislabel_opts=opts.LabelOpts(formatter="{value}"),
            )
        )
        .add_xaxis(data.index.astype(str).tolist())
        .add_yaxis(
            data.columns[0], 
            data.iloc[:, 0].tolist(),
            z=0
        )
        .set_series_opts(label_opts=opts.LabelOpts(is_show=False))
    )

    cumsum_line = (
        Line()
        .add_xaxis(data.index.astype(str).tolist())
        .add_yaxis(
            data.columns[1], 
            data.iloc[:, 1].tolist(),
            yaxis_index=1,
            linestyle_opts=opts.LineStyleOpts(width=1),
            markline_opts=opts.MarkLineOpts(
                data=[opts.MarkLineItem(y=0.8)],
                linestyle_opts=opts.LineStyleOpts(
                    width=1,
                    type_='dashed',
                    opacity=0.8,
                ),
                symbol='none',
                label_opts=opts.LabelOpts(is_show=False),
            ),
            z=1
        )
        .set_series_opts(label_opts=opts.LabelOpts(is_show=False))
    )
    
    return bar.overlap(cumsum_line).render_notebook() if is_show else bar.overlap(cumsum_line)

In [37]:
purchase_vc = user_gb[user_gb['purchase'] > 0]['purchase'].value_counts().sort_index()
purchase_vc.index = purchase_vc.index.astype(int)
purchase_vc = pd.DataFrame(purchase_vc).rename(columns={'purchase': '客户数'})
# purchase_vc
draw_pareto(purchase_vc, tittle='购买次数分布', x_name='购买次数')

现象：
- 当季最大客户购买次数高达 417。
- 购买次数大于 3（复购次数 > 2）的客户占 78%。

结论：成为客户后复购频率较高。

建议：
- 保持客户粘性。
- 促进用户转为客户，提升老客户比例。

### 复购间隔天数分布

In [38]:
user_purchase = data[data['purchase'] == 1][['user_id', 'date']].drop_duplicates().reset_index(drop=True)
user_merge = pd.merge(left=user_purchase, right=user_purchase, how='left', on='user_id')
user_merge = user_merge[user_merge['date_y'] > user_merge['date_x']]
user_repurchase = user_merge.groupby(['user_id', 'date_x']).first()
user_repurchase = user_repurchase.reset_index()
user_repurchase['ndays'] = user_repurchase['date_y'] - user_repurchase['date_x']
# user_repurchase
repurchase_days = user_repurchase['ndays'].value_counts().sort_index()
repurchase_days = pd.DataFrame(repurchase_days).rename(columns={'ndays': '复购次数'})
# repurchase_days
draw_pareto(repurchase_days, tittle='客户复购间隔天数分布', x_name='客户复购间隔天数')

现象：整体上，随着购后间隔天数增大，发生复购的次数越来越少。

**结论：购后的间隔天数越小，复购的概率越高。**

建议：在客户发生购买行为后的几天里，推荐与购买商品相关的其它商品。

### 消费金额分布

In [39]:
turnovers_vc = pd.cut(
    user_gb[user_gb['turnovers'] > 0]['turnovers'],
    bins=[i for i in range(0, int(np.max(user_gb['turnovers'])) + 50, 50)],
).value_counts().sort_index()
turnovers_vc = pd.DataFrame(turnovers_vc).rename(columns={'turnovers': '用户数'})
# turnovers_vc
draw_pareto(pd.DataFrame(turnovers_vc), x_name='消费金额', tittle='消费金额分布')

现象：
- 71% 的客户消费金额在 (0, 50] 的区间内。
- 90% 的客户消费金额在 (0, 100] 的区间内。
- 客户消费金额基本 < 400，消费金额大于 400 的客户占比小于 1%。

建议：发掘客户潜力，培养高端客户，提高客户消费金额。

### 客户分类

#### 传统 RFM 模型

构建 RFM 模型，对客户（本季度有过消费行为的用户）进行分层，方便跟踪各层用户的特征及变化情况，以针对性运营提高效率。

目标指标：
- Recency: 最近一次消费时间间隔
- Frequency：消费频率
- Monetary：消费金额

划分方式：以均值作为划分标准，将 RFM 三个指标数据转化为高、低两个等级变量，根据用户每项指标等级，划分为 8 类具不同消费特征的群体，由此直观反映各类客户的价值和重要性，并可根据相应特征采取针对性策略，可操作性强。

|R|F|M|划分群体类型|特点|
|-|-|-|-|-|
|高|高|高|重要价值客户|高端客户，各项指标都优于平均，质量最优|
|低|高|高|重要保持客户|最近购买日期早于平均最近购买日期，存在流失的可能性|
|高|低|高|重要发展客户|消费次数低于均值，用户粘性不高|
|低|低|高|重要挽留客户|消费次数低于均值和最近购买日期早于平均最近购买日期，流失风险大|
|高|高|低|一般价值客户|近期消费、频率高于平均，但消费金额低于平均|
|低|高|低|一般保持客户|频率高于平均、消费金额低于平均，近期无消费，存在流失的可能性|
|高|低|低|一般发展客户|近期消费，频率和消费金额均低于平均，可能是试水新用户或粘性不高的老用户|
|低|低|低|一般挽留客户|各项指标都低于平均|

注：由于消费金额 M 是衡量用户价值更重要的属性，根据消费高金额 M 的高低又划分为重要客户和一般客户。

In [2]:
# 统计消费用户 rfm 指标
user_rfm = pd.DataFrame(data['user_id'].unique(), columns=['user_id'])
user_recently = pd.DataFrame(data[data['purchase'] == 1].groupby('user_id', as_index=False)['date'].apply(lambda x: x.max()))
first_date = data['date'].min()
user_recently['date'] = (user_recently['date'] - first_date).dt.days
user_rfm = pd.merge(left=user_rfm, right=user_recently, how='left', on=['user_id'])
user_frequency = data[['user_id', 'purchase']].groupby('user_id').sum()
user_rfm = pd.merge(left=user_rfm, right=user_frequency, how='left', on=['user_id'])
user_monetary = data[['user_id', 'turnovers']].groupby('user_id').sum()
user_rfm = pd.merge(left=user_rfm, right=user_monetary, how='left', on=['user_id'])
user_rfm.fillna(0, inplace=True)
user_rfm.set_index('user_id', inplace=True)
user_rfm.columns = ['Recently','Frequency', 'Monetary']
# user_rfm

In [5]:
consumer_rfm = user_rfm[user_rfm['Frequency'] > 0]

# 划分 rfm 等级变量
rfm_divide = {col: consumer_rfm[col].mean() for col in consumer_rfm.columns}  # 取 rfm 中位数
consumer_rfm = pd.DataFrame(consumer_rfm)

for col in consumer_rfm:
    '''等级变量'''
    consumer_rfm[f"_{col}"] = consumer_rfm[col].apply(lambda x: 1 if x > rfm_divide[col] else 0)

# 根据 rfm 等级变量打标签
consumer_rfm['code'] = 100 * consumer_rfm['_Recently'] + 10 * consumer_rfm['_Frequency'] + consumer_rfm['_Monetary']  # 编码
rfm_code2label = {
    111: '重要价值客户',
    11: '重要保持客户',
    101: '重要发展客户',
    1: '重要挽留客户',
    110: '一般价值客户',
    10: '一般保持客户',
    100: '一般发展客户',
    0: '一般挽留客户',
}
consumer_rfm['label'] = consumer_rfm['code'].apply(lambda x: rfm_code2label[x])  # 编码转标签

# 3. 分层效果检验
# 3.1 各类用户占比
labels_order = ['重要价值客户', '重要保持客户', '重要发展客户', '重要挽留客户', '一般价值客户', '一般保持客户', '一般发展客户', '一般挽留客户']
label_vc = consumer_rfm['label'].value_counts().reindex(labels_order)

rfm_pie = (
    Pie()
    .add(
        '',
        [list(z) for z in zip(label_vc.index.tolist(), label_vc.values.tolist())],
        radius=[80, 180],
        rosetype="radius", #玫瑰饼图
    )
    .set_global_opts(
        title_opts=opts.TitleOpts('RFM 分层效果'),
        legend_opts=opts.LegendOpts(is_show=False)
    )
    .set_series_opts(label_opts=opts.LabelOpts(formatter='{b}: {d}%'))
)

# 3.2 各类用户 rfm 雷达图
label_mean = consumer_rfm.groupby('label')[['Recently', 'Frequency', 'Monetary']].mean()
label_mean = label_mean.reindex(labels_order)
colors = ['#D53A35', '#2F4554', '#61A0A8', '#D48265', '#749F83', '#CA8622', '#BDA29A', '#6E7074']
# 数据做归一化处理
label_mean_mms = MinMaxScaler().fit_transform(label_mean)

label_mean_radar = (
    Radar()
    .add_schema(
        schema=[opts.RadarIndicatorItem(name=col) for col in label_mean.columns]
    )
    .set_global_opts(
        title_opts=opts.TitleOpts("各类客户 RFM 雷达图"),
        legend_opts=opts.LegendOpts(pos_bottom='15%'),
    )
)

for i, v in enumerate(label_mean.index):
    label_mean_radar.add(v, [label_mean.loc[v].tolist()], color=colors[i])
    label_mean_radar.set_series_opts(label_opts=opts.LabelOpts(is_show=False))
    
label_monetary = pd.DataFrame(consumer_rfm.groupby('label')['Monetary'].sum()).reindex(labels_order)
label_monetary.columns = ['消费金额']
pareto = draw_pareto(label_monetary, x_name='客户类型', is_show=False)
    
rfm_page = (
    Page()
    .add(rfm_pie, label_mean_radar, pareto)
)
rfm_page.render_notebook()

现象：
- 重要客户占比约 30%，贡献整体销售额的 66%。其中，重要价值客户在整体客户中占比 13.94%，是第三大客户群体，贡献整体销售额的 33%。
- 一般客户占比约 70%，贡献整体销售额的 34%。其中，一般发展客户（32.51%）和一般挽留客户（29.51%）在整体客户中占比明显高于其它群体，是最大的两个客户群体。

**结论：客户质量两极分化，一般客户基数较大有待开发。**

建议：
- 经营好重要客户，保持重要客户的粘性。例如：
    - '重要价值客户'： VIP 优待 -- 提供更优质的售时售后服务，或进行不定期福利回馈。
    - '重要保持客户'：及时唤醒 -- 采取措施重新唤醒，促进客户复购。若存有联系方式，而可通过固定周期未购买进行消息通知。
    - '重要发展客户' 和 '重要挽留客户'：提高客户粘性 -- 增加对这类客户的曝光量和吸引力。前者可以及时推送优惠活动，后者可以通过调研客户消费需求和满意度调查，对能够改进的地方进行完善。
- 充分开发一般客户，使其向更高层客户群体次转化。可以针对这部分客户的进行消费需求调研和满意度调查，对能够改进的地方进行完善。

#### KMeans 聚类

In [67]:
consumer_kmeans = user_rfm[user_rfm['Frequency'] > 0]
consumer_ss = MinMaxScaler().fit_transform(consumer_kmeans)

# Calinski-Harabasz Index 评估自动取最高值的 K 为类数
best_k, best_calinski = 0, 0
for k in range(2, 9):
    y_pred = KMeans(n_clusters=k, random_state=1234).fit_predict(consumer_ss)
    calinski = metrics.calinski_harabasz_score(consumer_ss, y_pred)
#     print(f"{k}: {calinski}.")
    if calinski > best_calinski:
        best_k, best_calinski = k, calinski

# 使用自选择最佳 K 进行聚类
best_model = KMeans(n_clusters=best_k, random_state=1234)
best_model.fit(consumer_ss)
consumer_kmeans['label'] = best_model.labels_
label_mean = consumer_kmeans.groupby('label').mean().sort_values(by=['Monetary', 'Frequency', 'Recently'], ascending=False)
label_mapper = dict(zip(label_mean.index, [chr(ord('A') + i) for i in range(label_mean.shape[0])]))
consumer_kmeans['label'] = consumer_kmeans['label'].map(label_mapper)
label_mean.index = [chr(ord('A') + i) for i in range(label_mean.shape[0])]
label_vc = consumer_kmeans['label'].value_counts().reindex(labels_order)
colors = ['#D53A35', '#2F4554', '#61A0A8', '#D48265', '#749F83', '#CA8622', '#CA8622', '#6E7074']

kmeans_pie = (
    Pie()
        .add(
        '',
        [list(z) for z in zip(label_vc.index.tolist(), label_vc.values.tolist())],
        radius=[100, 180]
    )
        .set_global_opts(
        title_opts=opts.TitleOpts('K-Means 分层效果'),
        legend_opts=opts.LegendOpts(is_show=False)
    )
        .set_series_opts(label_opts=opts.LabelOpts(formatter='{b}: {d}%'))
)


label_mean_mms = MinMaxScaler().fit_transform(label_mean)
label_mean_radar = (
    Radar()
    .add_schema(
        schema=[opts.RadarIndicatorItem(name=col) for col in label_mean.columns]
    )
    .set_global_opts(
        title_opts=opts.TitleOpts("各类用户雷达图"),
        legend_opts=opts.LegendOpts(
        pos_right='0%',
            pos_top='middle',
            orient='vertical',
        ),
    )
)

for i, v in enumerate(label_mean.index):
    label_mean_radar.add(v, [label_mean.loc[v].tolist()], color=colors[i])
    label_mean_radar.set_series_opts(label_opts=opts.LabelOpts(is_show=False))
    
label_monetary = pd.DataFrame(consumer_kmeans.groupby('label')['Monetary'].sum())
label_monetary.columns = ['贡献销售额']
pareto = draw_pareto(label_monetary, is_show=False)

kmeans_page = (
    Page()
    .add(kmeans_pie, label_mean_radar, pareto)
)
kmeans_page.render_notebook()

现象：使用 KMeans 基于 Calinski-Harbasz Score 自动选择最佳聚类数为 6 类。分层属性中，各类用户“最近购买时间”的差异较大，而“购买金额”和“购买频次”差异较小。

**结论：对“最近购买时间”较远的用户群体应及时进行召回。**

## 品牌分析

### 品牌商品数量

In [41]:
brand_gb_fun = {
    'view': np.sum, 
    'cart': np.sum, 
    'remove_from_cart': np.sum, 
    'purchase': np.sum, 
    'turnovers': np.sum,
    'price': [np.median, np.mean],  # [np.min, np.max, np.median, np.mean]
    'product_id': lambda x: x.nunique()
}
brand_gb = data.groupby('brand').agg(brand_gb_fun)
brand_gb.columns = ['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers', 'median_price', 'mean_price', 'products_n']
brand_gb = brand_gb.drop(index=['MISSING'])  # 删除缺失品牌
brand_gb['avg_view'] = np.round(brand_gb['view'] / brand_gb['products_n'], 2)
brand_gb['avg_cart'] = np.round(brand_gb['cart'] / brand_gb['products_n'], 2)
brand_gb['avg_purchase'] = np.round(brand_gb['purchase'] / brand_gb['products_n'], 2)
brand_gb['turnovers/purchase'] = np.round(brand_gb['turnovers'] / brand_gb['purchase'], 2)
brand_gb['turnovers/purchase'] = brand_gb['turnovers/purchase'].fillna(0)

In [42]:
brandn_wc_data = [(brand, n) for brand, n in zip(brand_gb.index.tolist(), brand_gb['products_n'].tolist())]
brand_wc = (
    WordCloud()
    .add('品牌商品数量', brandn_wc_data)
    .set_global_opts(
        title_opts=opts.TitleOpts(
            title='品牌商品数量',
            subtitle='注：已过滤品配缺失商品',
        ),
    )
)
brand_wc.render_notebook()

现象：商品数量大于 1000 的品牌有 masura(1562)、irisk(1325)、runail(1274) ，数量级明显大于其它品牌（blusky(893) 商品数量排名第四）。

### 品牌成交额

In [43]:
brand_turnovers = brand_gb[brand_gb['turnovers'] > 0][['turnovers']].sort_values(by=['turnovers'], ascending=False)
brand_turnovers.columns = ['品牌成交额']
brand_turnovers['rank'] = brand_turnovers['品牌成交额'].rank(method='first', ascending=False)
brand_turnovers['rank'] = brand_turnovers['rank'].astype(int)
brand_turnovers = brand_turnovers.reset_index()
brand_turnovers['brand'] = brand_turnovers['rank'].astype(str) + '.' + brand_turnovers['brand'].astype(str)
brand_turnovers = brand_turnovers.set_index('brand')
del brand_turnovers['rank']
# brand_turnovers
draw_pareto(brand_turnovers, tittle='品牌成交额帕累托图')

现象：成交额排名前 35（占品牌总数的 15%）的品牌贡献了 80% 的成交额。

### 品牌矩阵分析

In [44]:
brand_info = brand_gb[['turnovers', 'products_n', 'avg_view', 'avg_purchase', 'turnovers/purchase']]
brand_gb_corr = brand_info.corr()
row_n, col_n = brand_gb_corr.values.shape
brand_gb_corr = np.round(brand_gb_corr.values, 2).tolist()
brand_corr = [[row, col, brand_gb_corr[row][col]] for row in range(row_n) for col in range(col_n)]
brand_corr_heatmap = (
    HeatMap()
    .add_xaxis(brand_info.columns.tolist())
    .add_yaxis(
        "",
        brand_info.columns.tolist(),
        brand_corr,
        label_opts=opts.LabelOpts(
            is_show=True, color="#fff",position='inside'
        ),
    )
    .set_global_opts(
        title_opts=opts.TitleOpts(title="品牌指标相关系数"),
        visualmap_opts=opts.VisualMapOpts(
            max_=np.max(brand_gb_corr),
            min_=np.round(np.min(brand_gb_corr) - 0.1, 1),
            pos_right='right',
            pos_bottom='10%',
            is_calculable=True,
            precision=2,
        ),
        xaxis_opts=opts.AxisOpts(
            type_="category",
            splitarea_opts=opts.SplitAreaOpts(
                is_show=True, areastyle_opts=opts.AreaStyleOpts(opacity=3)
            ),
        ),
        yaxis_opts=opts.AxisOpts(
            type_="category",
            splitarea_opts=opts.SplitAreaOpts(
                is_show=True, areastyle_opts=opts.AreaStyleOpts(opacity=1)
            ),
        ),
    )
)
brand_corr_heatmap.render_notebook()

可见，强相关（r > 0.5）的属性有：品牌成交额与品牌商品数量；品牌平均浏览量和品牌成交均价。

In [45]:
name = brand_gb.index.tolist()
bvpt = brand_gb[['view', 'purchase', 'turnovers', 'median_price', 'mean_price',]]

symbol_js = """
    function (data) {return Math.sqrt(data[3] / 88);}
    """

scatter_formatter_js = '''
    function(params){
        return '品牌：' + params.value[2]
        + '<br/>流量：' + params.value[0]
        + '<br/>销量：' + params.value[1]
        + '<br/>成交额：' + params.value[3]
        + '<br/>商品价格中位数：' + params.value[4]
        + '<br/>商品均价：' + params.value[5];
    }
    '''

pt_scatter = (
    Scatter()
    .add_xaxis(bvpt['view'])
    .add_yaxis(
        "",
        [list(z) for z in zip(bvpt['purchase'], name, bvpt['turnovers'], bvpt['median_price'], bvpt['mean_price'],)],
        symbol_size=JsCode(symbol_js),
        label_opts=opts.LabelOpts(
            is_show=False,
            formatter=JsCode(
                "function(params){return params.value[3];}"
            ),
        ),
    )
    .set_global_opts(
        title_opts=opts.TitleOpts(
            title="品牌流量与销量矩阵",
            subtitle='注：点的大小与成交额正相关；虚线为均值。',
        ),
        xaxis_opts=opts.AxisOpts(name='流量', type_="value", name_location='middle', name_rotate=0, name_gap=30),
        yaxis_opts=opts.AxisOpts(name='销量', type_="value", name_location='middle', name_rotate=0, name_gap=50),
        tooltip_opts=opts.TooltipOpts(
            formatter=JsCode(scatter_formatter_js),
        ),
        datazoom_opts=opts.DataZoomOpts(
            is_show=True,
            orient='vertical',
            range_start=0,
            range_end=100,
        ),
    )
    .set_series_opts(
        markline_opts=opts.MarkLineOpts(
            is_silent=True,
            data=[
                opts.MarkLineItem(
                    name="流量均值",
                    type_='value',
                    x=round(np.mean(bvpt['view']), 0),
                ),
                opts.MarkLineItem(
                    name="销量均值",
                    type_='value',
                    y=round(np.mean(bvpt['purchase']), 0),
                ),
            ],
            symbol='none',
            label_opts=opts.LabelOpts(
                is_show=False,
            ),
        ),
    )
)
pt_scatter.render_notebook()

In [46]:
level = {
    11: '高流量-高销量',
    10: '高流量-低销量',
    1: '低流量-高销量',
    0: '低流量-低销量',
}

for c in ['view', 'purchase']:
    bvpt[f"{c}_level"] = (bvpt[c] > bvpt[c].mean()).astype(int)
bvpt['level'] = bvpt['view_level'] * 10 + bvpt['purchase_level']
bvpt['level'] = bvpt['level'].map(level)
# bvpt
level_vc = bvpt.groupby('level').agg({'level':np.size, 'purchase':np.sum, 'turnovers':np.sum})
level_vc['品牌成交均价'] = np.round(level_vc['turnovers'] / level_vc['purchase'], 2)
del level_vc['purchase'], level_vc['turnovers']
level_vc = level_vc.rename(columns={'level': '品牌数量'}).reindex(level.values()).reset_index()
# level_vc

level_pie = (
    Pie()
    .add(
        '',
        [list(z) for z in zip(level_vc['level'], level_vc['品牌数量'])],
        radius=[80, 180],
    )
    .set_global_opts(
        title_opts=opts.TitleOpts(
            title='品牌矩阵四象限分布',
            pos_left='center',
        ),
        legend_opts=opts.LegendOpts(is_show=False),
    )
    .set_series_opts(label_opts=opts.LabelOpts(formatter='{b}: {d}%'))
)
level_pie.render_notebook()

In [47]:
level_table = Table()
headers = level_vc.columns.tolist()
rows = [list(z) for z in zip(level_vc['level'], level_vc['品牌数量'], level_vc['品牌成交均价'])]
level_table.add(headers, rows)
level_table.render_notebook()

level,品牌数量,品牌成交均价
高流量-高销量,38,4.34
高流量-低销量,10,33.78
低流量-高销量,1,5.82
低流量-低销量,195,7.19


根据品牌浏览量与成交量的**均值**将 216 个品牌划分至四象限：

   - 高流量，高销量：38 个品牌，占比 15.57%，平均成交价格 4.34 为四象限中最低
   - 高流量，低销量：10 个品牌，占比 4.1%，平均成交价格 33.78 为四象限中最高
   - 低流量，高销量：仅 1 个品牌，占比 0.41%
   - 低流量，低销量：195 个品牌，占比 79.92%

现象：
1. 高流量品牌占整体约 20%，其中高销量占整体约 16%， 低销量占整体约 14%，前者成交均价为 4.34，后者成交均价为 33.78，约为前者 8 倍
2. 低流量品牌占整体约 80%，其中高销量仅占整体约 0.4%，低销量占整体约 90%，成交均价分别为 5.82、7.19，二者差距较小，且略大于高流量-高销量品牌

**结论：高流量不一定高销量，但低流量几乎必然低销量。在高流量的情况下，低价商品销量更高，但低价并非高销量的必然前提。**

建议：
1. 尽可能增加品牌的曝光度并转化为有效流量，提高流量是提高销量的必然前提。
2. 提高流量的同时，合理控制价格，以期成交额的最大化。